### PR TITLE
LibCore+RS: Make RequestServer (spin) faster

### PR DESCRIPTION
**Pull request by alimpfard**
See individual commits; perf measurements coming soon :tm:

### CODE REVIEW

File: Userland/Services/RequestServer/ConnectionCache.cpp [hunk 1]
```diff
@@ -26,8 +26,17 @@ void request_did_finish(URL::URL const& url, Core::Socket const* socket)
 
     ConnectionKey partial_key { url.serialized_host().release_value_but_fixme_should_propagate_errors().to_byte_string(), url.port_or_default() };
     auto fire_off_next_job = [&](auto& cache) {
-        auto it = find_if(cache.begin(), cache.end(), [&](auto& connection) { return connection.key.hostname == partial_key.hostname && connection.key.port == partial_key.port; });
-        if (it == cache.end()) {
+        auto [it, end] = cache.with_locked([&](auto& cache) {
+            struct Result {
+                decltype(cache.begin()) it;
```
Comment (ADKaster) [inline Userland/Services/RequestServer/ConnectionCache.cpp:31]: surely we have a typedef for ::Iterator :thinkies:

File: Userland/Services/RequestServer/ConnectionCache.h [hunk 2]
```diff
@@ -185,84 +188,86 @@ decltype(auto) get_or_create_connection(auto& cache, URL::URL const& url, auto j
...
-    if (failed_to_find_a_socket && sockets_for_url.size() < ConnectionCache::MaxConcurrentConnectionsPerURL) {
-        using ConnectionType = RemoveCVReference<decltype(*cache.begin()->value->at(0))>;
-        auto connection_result = proxy.tunnel<typename ConnectionType::SocketType, typename ConnectionType::StorageType>(url);
-        if (connection_result.is_error()) {
-            dbgln("ConnectionCache: Connection to {} failed: {}", url, connection_result.error());
-            Core::deferred_invoke([job] {
-                job->fail(Core::NetworkJob::Error::ConnectionFailed);
-            });
-            return ReturnType { nullptr };
+    return cache.with_locked([&] (auto&) -> ReturnType {
```
Comment (ADKaster) [inline Userland/Services/RequestServer/ConnectionCache.h:191]: this is an absolutely massive lambda. Can we yoink its internals into a function that verifies that the cache is locked?

File: Userland/Services/RequestServer/ConnectionFromClient.cpp [hunk 3]
```diff
@@ -26,8 +26,140 @@ static IDAllocator s_client_ids;
...
+
+    URL::URL m_url;
+    inline static HashMap<URL::URL, WeakPtr<Job>> s_jobs {};
+};
+
+ErrorOr<void> ConnectionFromClient::worker_main()
+{
+    Core::EventLoop thread_loop;
+
+    NonnullRefPtr<Core::Timer> timer = Core::Timer::create_repeating(100, [&] {
```
Comment (ADKaster) [inline Userland/Services/RequestServer/ConnectionFromClient.cpp:88]: same comment here about massive lambdas :S

File: Userland/Services/RequestServer/ConnectionFromClient.cpp [hunk 4]
```diff
@@ -26,8 +26,140 @@ static IDAllocator s_client_ids;
...
+        // Don't restart this function while we're processing.
+        timer->stop();
+        ScopeGuard guard([&] {
+            timer->start();
+        });
+
+        do {
+            if (!m_work_queue.is_valid())
```
Comment (ADKaster) [inline Userland/Services/RequestServer/ConnectionFromClient.cpp:96]: what's an "invalid" work queue :thinking:

File: Userland/Services/RequestServer/ConnectionFromClient.cpp [hunk 5]
```diff
@@ -62,6 +194,14 @@ Messages::RequestServer::ConnectNewClientResponse ConnectionFromClient::connect_
     return IPC::File::adopt_fd(socket_fds[1]);
 }
 
+void ConnectionFromClient::enqueue(Work work)
+{
+    Core::deferred_invoke_if([this, work = move(work)] {
```
Comment (ADKaster) [inline Userland/Services/RequestServer/ConnectionFromClient.cpp:199]: so wait, here we just shuffle the jobs > the 256th around in vectors in the event loop until they can be pushed into the work queue? How is that better than allowing the work queue to expand, or storing them on a secondary queue in this object?

Comment (ADKaster) [reply Userland/Services/RequestServer/ConnectionFromClient.cpp:199]: e.g. a check to say 

enqueue: if the extra queue has items or we fail to enqueue to the work queue, push to that secondary queue w/it locked. 
worker_main: once queue is empty, try to move items from secondary queue to work queue w/secondary queue locked.

File: Userland/Services/RequestServer/ConnectionCache.h (hunk 2 above)
Comment (alimpfard) [reply Userland/Services/RequestServer/ConnectionCache.h:191]: ...how do you verify anything with a MutexProtected?
will move it out into a different function tho

File: Userland/Services/RequestServer/ConnectionFromClient.cpp (hunk 4 above)
Comment (alimpfard) [reply Userland/Services/RequestServer/ConnectionFromClient.cpp:96]: deleted on the other side, it's...weird.

File: Userland/Services/RequestServer/ConnectionFromClient.cpp (hunk 5 above)
Comment (alimpfard) [reply Userland/Services/RequestServer/ConnectionFromClient.cpp:199]: we don't have a threadsafe expanding queue, and I don't want to invent one.
I'd rather let the queue grow as well, but no infra

Comment (alimpfard) [reply Userland/Services/RequestServer/ConnectionFromClient.cpp:199]: re locked secondary queue, no point in using SSPCQ then, might as well just have a single locked queue.

File: Userland/Services/RequestServer/ConnectionFromClient.cpp (hunk 3 above)
Comment (alimpfard) [reply Userland/Services/RequestServer/ConnectionFromClient.cpp]: Moved the lambda.

File: Userland/Services/RequestServer/ConnectionCache.h (hunk 2 above)
Comment (alimpfard) [reply Userland/Services/RequestServer/ConnectionCache.h]: Moved to a new function.

**Comment by alimpfard** [comment]
I don't like where this is going.

**Pull request closed without merging**